limits.c

```diff
@@ -330,81 +335,150 @@ void limits_soft_check(float *target)
   
-  uint8_t axislock = AXISLOCKSERVO;
-  servo_rate = settings.homing_seek_rate[X_AXIS]/20.0; // TODO: make servo_rate into macro for servoing
+  uint8_t axislock = 1 << Z_AXIS;
   plan_reset();
 
   // set target for moving axes based on direction
   target[X_AXIS] = 0;
   target[Y_AXIS] = 0;
-  target[Z_AXIS] = travel_servo;
+  target[Z_AXIS] = travel;
   target[C_AXIS] = 0;
 
   // Perform homing cycle. Planner buffer should be empty, as required to initiate the homing cycle.
-  plan_buffer_line(target, servo_rate, false, LINENUMBER_SPECIAL_SERVO|(servo_line_number*4+LINEMASK_ON_EDGE)); 
+  plan_buffer_line(target, servo_rate, false, LINENUMBER_EMPTY_BLOCK);
+  
+  // Axislock bit is high if axis is homing, so we only enable checking on moving axes.
+  // Expect 0 on approach (stop when 1). vice versa for pulloff
 
-  // axislock bit is high if axis is homing, so we only enable checking on moving axes.
-  limits_enable(axislock,~approach);  //expect 0 on approach (stop when 1). vice versa for pulloff
-  limits.isservoing = 1; // tell system we are servoing
+  limits_enable(axislock,~approach);  
+  // Set the isservoing flag
+  limits.isservoing = 1;
 
   st_prep_buffer(); // Prep and fill segment buffer from newly planned block.
   st_wake_up(); // Initiate motion
 
+
+}
+
+// Move gripper to limits.bump_grip_force.
+void limits_force_servo(){
+  // Block if system reset has been issued.
+  if (sys.abort) {
+    return;
+  } 
+
+  // Do not look for gaps in magazine on carousel
+  limits.mag_gap_check = 0;  
+
+  float servo_rate = settings.homing_seek_rate[Z_AXIS];
+
+  // Pause the callback function that updates ADC
+  // values when not force servoing.
+  signals.pause = 1;
+
+  // The initial seek flag allows the gripper motor to move close
+  // to the desired force and then decrease the servo rate to seek
+  // the desired force more accurately 
+  uint8_t initial_seek = 1; 
+
+  // Delays are used to compensate for the settling time
+  // of the load cell ADC reading
+  uint16_t delay = 0; 
+
+ // Keep track of completed cycles
+  uint8_t cycles = 0; 
+
+  // Perform force servoing cyles
   do {
-    st_prep_buffer(); // Check and prep segment buffer. NOTE: Should take no longer than 200us.
-    // Check only for user reset. Keyme: fixed to allow protocol_execute_runtime() in this loop.
-    protocol_execute_runtime();
-    if (SYS_EXEC & EXEC_RESET) {
-      protocol_execute_runtime();
-      return;
+    if (initial_seek) {
+      // Initially, seek at a fast rate
+      limits_plan_force_servo(servo_rate);
+
+      // Clear the initial seek flag
+      initial_seek = 0;  
+
+      // After servoing at a fast rate, 
+      // delay for LONG_FORCE_SERVO_DELAY ms 
+      delay = LONG_FORCE_SERVO_DELAY;
+
+    } else {
+      // Step at the slowest speed possible
+      limits_plan_force_servo(1);
+
+      // After servoing at a fast rate, 
+      // delay for SHORT_FORCE_SERVO_DELAY ms
+      delay = SHORT_FORCE_SERVO_DELAY;
+    
```

Comment:
It seems like you could avoid this if/else statement here if you performed the 'initial_seek' true branch before entering the outer do loop, and then performed the false branch at the end of the outer do loop.